jetty-util/src/main/java/org/eclipse/jetty/util/TypeUtil.java

```diff
@@ -751,4 +755,89 @@ public T next()
+     * it warns and continues iterating through the service loader.
+     * @param <T> The class of the service type.
+     * @param serviceLoader The service loader to use.
+     * @return a list of the loaded service providers.
+     * @throws ServiceConfigurationError If the number of errors exceeds {@link #MAX_ERRORS}
+     */
+    public static <T> List<T> loadAll(ServiceLoader<T> serviceLoader)
+    {
+        List<T> list = new ArrayList<>();
+        Iterator<T> iterator = serviceLoader.iterator();
+
+        int errors = 0;
+        while (true)
+        {
+            try
+            {
+                if (!iterator.hasNext())
+                    break;
+                list.add(iterator.next());
+            }
+            catch (ServiceConfigurationError e)
+            {
+                LOG.warn(e);
+                if (++errors >= MAX_ERRORS)
+                    throw e;
+            }
+        }
+
+        return list;
+    }
+
+    /**
+     * Uses the {@link ServiceLoader} to get the first available service provider.
+     * If loading a service type throws {@link ServiceConfigurationError},
+     * it warns and continues iterating through the service loader until one is found.
+     * @param <T> The class of the service type.
+     * @param serviceLoader The service loader to use.
+     * @return an instance of a service provider, or null if none could be found.
+     * @throws ServiceConfigurationError If the number of errors exceeds {@link #MAX_ERRORS}
+     */
+    public static <T> T loadFirst(ServiceLoader<T> serviceLoader)
+    {
+        return loadFirst(serviceLoader, t -> true);
+    }
+
+    /**
+     * Uses the {@link ServiceLoader} to get the first service provider to match the provided predicate.
+     * If loading a service type throws {@link ServiceConfigurationError},
+     * it warns and continues iterating through the service loader until one is found.
+     * @param <T> The class of the service type.
+     * @param serviceLoader The service loader to use.
+     * @param predicate The predicate used to match a service provider.
+     * @return an instance of a service provider, or null if none could be found.
+     * @throws ServiceConfigurationError If the number of errors exceeds {@link #MAX_ERRORS}
+     */
+    public static <T> T loadFirst(ServiceLoader<T> serviceLoader, Predicate<T> predicate)
+    {
+        Iterator<T> iterator = serviceLoader.iterator();
+
+        int errors = 0;
+        while (true)
+        {
+            try
+            {
+                if (!iterator.hasNext())
+                    break;
+
+                T t = iterator.next();
+                if (predicate.test(t))
+                    return t;
+            }
+            catch (ServiceConfigurationError e)
+            {
+                LOG.warn(e);
+                if (++errors >= MAX_ERRORS)
+                    throw e;
+            }
+        }
+
+        return null;
```

Comment:
No, should return `Optional<T>`, not `null`.

But frankly I would just have 1 method that returns a `Stream`, and the caller can easily load them all into a `List`, or load the first, etc. using the `Stream` APIs.

Reply:
Optional only really works if Streams are being used.   I think it is a mistake to mix styles.

Reply:
I don't think Optional is really necessary as it only is advantageous if you are using streams.
So I think it is a Stream API or as is, no need to mix styles.

Reply:
`Optional` is orthogonal to `Stream`. 
The signature with `Optional` makes clear that you may or may not get a service. Without the `Optional` you have to either read the source code or the javadocs.

Reply:
Hmmm I can't actually find any usages of loadFirst, so I hate methods that are not actually used. 
Are there any usages that can be converted to use loadFirst?  if not, then perhaps just doing a single more flexible Stream API is best.